lib/3scale/backend/storage_async/client.rb

```diff
@@ -41,131 +34,17 @@ def instance(reset = false)
         end
 
         def initialize(opts)
-          host, port = opts[:url].match(HOST_PORT_REGEX).captures if opts[:url]
-          host ||= DEFAULT_HOST
-          port ||= DEFAULT_PORT
-
-          endpoint = Async::IO::Endpoint.tcp(host, port)
-          @redis_async = Async::Redis::Client.new(
-            endpoint, limit: opts[:max_connections]
-          )
-          @building_pipeline = false
-        end
-
-        # Now we are going to define the methods to run redis commands
-        # following the interface of the redis-rb lib.
-        #
-        # These are the different cases:
-        #   1) Methods that can be called directly. For example SET:
-        #      @redis_async.call('SET', some_key)
-        #   2) Methods that need to be "boolified". These are methods for which
-        #      redis-rb returns a boolean, but redis just returns an integer.
-        #      For example, Redis returns 0 or 1 for the EXISTS command, but
-        #      redis-rb transforms that into a boolean.
-        #   3) There are a few methods that need to be treated differently and
-        #      do not fit in any of the previous categories. For example, SSCAN
-        #      which accepts a hash of options in redis-rb.
-        #
-        # All of this might be simplified a bit in the future using the
-        # "methods" in async-redis
-        # https://github.com/socketry/async-redis/tree/master/lib/async/redis/methods
-        # but there are some commands missing, so for now, that's not an option.
-
-        METHODS_TO_BE_CALLED_DIRECTLY = [
-          :del,
-          :expire,
-          :expireat,
-          :flushdb,
-          :get,
-          :hset,
-          :hmget,
-          :incr,
-          :incrby,
-          :keys,
-          :llen,
-          :lpop,
-          :lpush,
-          :lrange,
-          :ltrim,
-          :mget,
-          :ping,
-          :rpush,
-          :scard,
-          :setex,
-          :smembers,
-          :sunion,
-          :ttl,
-          :zcard,
-          :zrangebyscore,
-          :zremrangebyscore,
-          :zrevrange
-        ].freeze
-        private_constant :METHODS_TO_BE_CALLED_DIRECTLY
-
-        METHODS_TO_BE_CALLED_DIRECTLY.each do |method|
-          define_method(method) do |*args|
-            @redis_async.call(method, *args.flatten)
-          end
+          @redis_async = Concurrent::ThreadLocalVar.new{ initialize_client(opts) }
```

Comment:
WE do not have a good reason to do this, do we? When this is running in production, it is not needed.

Reply:
I think so, we've confirmed that production will run in single-thread mode so this won't be needed. I'd like to remove it but first I need a way to fix `spec/integration/worker_async_spec.rb`. The test is wrong and blocks the  test suite if I remove the `Concurrent::ThreadLocalVar` trick.

Reply:
I also guess that there is some issue in the tests.

Reply:
I think we should keep this unless it proofs to be a performance issue. This makes the class safer and it may help with future improvements of the deployment.

Reply:
IMO we should remove  that. We want to make sure the system fails if, for whatever reason, that `redis_async` instance is shared between threads.

Reply:
Does it fail fast presently? I thought only in presence of traffic it starts complaining.

This change maybe resolves the thread safety issue. And we need to validate changes of deployment and code with good amount of traffic either way. So I don't think this would cause any additional risk.

Reply:
I pushed some changes (https://github.com/3scale/apisonator/pull/350/commits/8c281104a65f4cdf3115e4d3ed67355cf0c9a90a) to remove this. While `Concurrent::ThreadLocalVar` has been very useful to understand what was happening, right now it's just unused temporary code. I prefer to get rid of it.

Also, I updated (https://github.com/3scale/apisonator/pull/350/commits/cdc25bb790695ea7e6f0aa03fd76282f161fc19f) the test to be sure it doesn't share the Redis client between threads. I have to say I'm still unsure about this test, specially because in the past I made similar changes to it and it failed, so I don't know why it worked today. It could be just because in the past I had less background and missed something.

Reply:
>  it should never be multi-threaded.
 
 If it yields performance gains, I see no reason not to become multi-threaded some day. For a class that is used globally (through `::instance`) I find it a limitation not to be thread safe. Alternatively `::instance` may return a different one per thread which might be even better.
 
Either way, I'm fine with whatever Joan decided to do.
 
 >  I'm still unsure about this test
 
 did you try increasing `n_reports`? Otherwise it is a hit and miss

Reply:
> If it yields performance gains, I see no reason not to become multi-threaded some day.

That would require, for N threads, N reactors and N redis-async clients. Currently the model is 1 pod -> 1 process -> 2 threads (1 for reading queue jobs, 1 for processing jobs)

Indeed, in the future, we can explore this. But currently the scaling approach is per pod. 1 pod consuming 1 CPU at max.

Reply:
> did you try increasing n_reports? Otherwise it is a hit and miss

I didn't try, I forgot the basics! 😥. But I have tried just now and it works fine. The tests pass with n_reports = 50, 500, 5000 and 50000. With no `Concurrent::ThreadLocalVar`. Always passes.

> That would require, for N threads, N reactors and N redis-async clients.

"If some day I have the time" I'd like to explore this, I think the main problem to implement multi-threading in apisonator is the cache memory implemented by memoization, which isn't thread-safe.

Reply:
> "If some day I have the time" I'd like to explore this, I think the main problem to implement multi-threading in apisonator is the cache memory implemented by memoization, which isn't thread-safe.

For bare metal with N cpus, this would make sense. In openshift with scaling based on pods, having one pod - 1 cpu is good enough to scale based on pods. We are currently working in adding [HPA](https://docs.openshift.com/container-platform/4.14/nodes/pods/nodes-pods-autoscaling.html) to the backend workers in https://issues.redhat.com//browse/THREESCALE-10500